// solid-ag-grid/dist/server.cjs
'use strict';

var web = require('solid-js/web');
var agGridCommunity = require('ag-grid-community');
var solidJs = require('solid-js');

// src/grid/agGridSolid.tsx

// src/grid/core/solidCompWrapper.tsx
var SolidCompWrapper = class {
  constructor(SolidCompClass, portalManager) {
    this.SolidCompClass = SolidCompClass;
    this.portalManager = portalManager;
  }
  init(props) {
    this.eGui = document.createElement("div");
    this.eGui.className = "ag-solid-container";
    this.portalInfo = {
      mount: this.eGui,
      SolidClass: this.SolidCompClass,
      props,
      ref: (instance) => {
        this.instance = instance;
      }
    };
    this.portalManager.addPortal(this.portalInfo);
  }
  destroy() {
    this.portalInfo && this.portalManager.removePortal(this.portalInfo);
  }
  getGui() {
    return this.eGui;
  }
  hasMethod(name) {
    return this.instance[name] != null;
  }
  getFrameworkComponentInstance() {
    return this.instance;
  }
  callMethod(name, args) {
    return this.instance[name].apply(this.instance, args);
  }
  addMethod(name, callback) {
    this[name] = callback;
  }
};

// src/grid/core/solidCompWrapperFactory.tsx
var SolidCompWrapperFactory = class extends agGridCommunity.BaseComponentWrapper {
  constructor(portalManager) {
    super();
    this.portalManager = portalManager;
  }
  createWrapper(SolidComponentClass) {
    return new SolidCompWrapper(SolidComponentClass, this.portalManager);
  }
};
var BeansContext = solidJs.createContext({});

// src/grid/core/utils.tsx
var classesList = (...list) => {
  const filtered = list.filter((s) => s != null && s !== "");
  return filtered.join(" ");
};
var CssClasses = class _CssClasses {
  classesMap = {};
  constructor(...initialClasses) {
    initialClasses.forEach((className) => {
      this.classesMap[className] = true;
    });
  }
  setClass(className, on) {
    const nothingHasChanged = !!this.classesMap[className] == on;
    if (nothingHasChanged) {
      return this;
    }
    const res = new _CssClasses();
    res.classesMap = {
      ...this.classesMap
    };
    res.classesMap[className] = on;
    return res;
  }
  toString() {
    const res = Object.keys(this.classesMap).filter((key) => this.classesMap[key]).join(" ");
    return res;
  }
};

// src/grid/cellRenderer/detailCellRenderer.tsx
var _tmpl$ = ["<div", ">", "</div>"];
var DetailCellRenderer = (props) => {
  const {
    ctrlsFactory,
    context,
    gridOptionsService,
    resizeObserverService,
    clientSideRowModel,
    serverSideRowModel
  } = solidJs.useContext(BeansContext);
  const [getCssClasses, setCssClasses] = solidJs.createSignal(new CssClasses());
  const [getGridCssClasses, setGridCssClasses] = solidJs.createSignal(new CssClasses());
  const [getDetailGridOptions, setDetailGridOptions] = solidJs.createSignal();
  const [getDetailRowData, setDetailRowData] = solidJs.createSignal();
  let ctrl;
  let eGuiRef;
  const getCssClassesStr = solidJs.createMemo(() => getCssClasses().toString() + " ag-details-row");
  const getGridCssClassesStr = solidJs.createMemo(() => getGridCssClasses().toString() + " ag-details-grid");
  props.ref(() => ({
    // force new instance when grid tries to refresh
    refresh() {
      return ctrl.refresh();
    }
  }));
  solidJs.onMount(() => {
    if (props.template && typeof props.template === "string") ;
    const compProxy = {
      addOrRemoveCssClass: (name, on) => setCssClasses(getCssClasses().setClass(name, on)),
      addOrRemoveDetailGridCssClass: (name, on) => setGridCssClasses(getGridCssClasses().setClass(name, on)),
      setDetailGrid: (gridOptions) => setDetailGridOptions(gridOptions),
      setRowData: (rowData) => setDetailRowData(rowData),
      getGui: () => eGuiRef
    };
    ctrl = ctrlsFactory.getInstance("detailCellRenderer");
    if (!ctrl) {
      return;
    }
    context.createBean(ctrl);
    ctrl.init(compProxy, props);
    let resizeObserverDestroyFunc;
    if (gridOptionsService.get("detailRowAutoHeight")) {
      const checkRowSizeFunc = () => {
        {
          return;
        }
      };
      resizeObserverDestroyFunc = resizeObserverService.observeResize(eGuiRef, checkRowSizeFunc);
    }
    return () => {
      context.destroyBean(ctrl);
      if (resizeObserverDestroyFunc) {
        resizeObserverDestroyFunc();
      }
    };
  });
  return web.ssr(_tmpl$, web.ssrAttribute("class", web.escape(getCssClassesStr(), true), false), getDetailGridOptions() && web.escape(web.createComponent(agGridSolid_default, web.mergeProps({
    get ["class"]() {
      return getGridCssClassesStr();
    }
  }, getDetailGridOptions, {
    get rowData() {
      return getDetailRowData();
    }
  }))));
};
var detailCellRenderer_default = DetailCellRenderer;
var JsUserComp = (p) => {
  const {
    context
  } = solidJs.useContext(BeansContext);
  const promise = p.compDetails.newAgStackInstance();
  if (!promise) {
    return [];
  }
  const comp = promise.resolveNow(null, (x) => x);
  if (!comp) {
    return [];
  }
  p.ref && p.ref(comp);
  const gui = comp.getGui();
  solidJs.onCleanup(() => {
    comp && context.destroyBean(comp);
    p.ref && p.ref(void 0);
  });
  return gui;
};
var jsUserComp_default = JsUserComp;
var SolidUserComp = (props) => {
  const SolidClass = props.compDetails.componentClass;
  solidJs.onCleanup(() => {
  });
  return web.createComponent(SolidClass, web.mergeProps(() => props.compDetails.params));
};
var solidUserComp_default = SolidUserComp;

// src/grid/userComps/userComp.tsx
var UserComp = (p) => {
  const showSolidComp = solidJs.createMemo(() => {
    const details = p.compDetails;
    if (!details) {
      return false;
    }
    return details.componentFromFramework;
  });
  const showJsComp = solidJs.createMemo(() => {
    const details = p.compDetails;
    if (!details) {
      return false;
    }
    return !details.componentFromFramework;
  });
  return [showSolidComp() && web.createComponent(solidUserComp_default, {
    get compDetails() {
      return p.compDetails;
    }
  }), showJsComp() && web.createComponent(jsUserComp_default, {
    get compDetails() {
      return p.compDetails;
    }
  })];
};
var userComp_default = UserComp;

// src/grid/cellRenderer/groupCellRenderer.tsx
var _tmpl$2 = ["<span", "></span>"];
var _tmpl$22 = ['<span class="ag-group-value">', "", "</span>"];
var _tmpl$3 = ['<span class="ag-group-child-count">', "</span>"];
var GroupCellRenderer = (props) => {
  const context = solidJs.useContext(BeansContext).context;
  let eGui;
  let eCheckboxRef;
  let eExpandedRef;
  let eContractedRef;
  let role = "gridcell";
  const [getInnerCompDetails, setInnerCompDetails] = solidJs.createSignal();
  const [getChildCount, setChildCount] = solidJs.createSignal();
  const [getValue, setValue] = solidJs.createSignal();
  const [getCssClasses, setCssClasses] = solidJs.createSignal(new CssClasses());
  const [getExpandedCssClasses, setExpandedCssClasses] = solidJs.createSignal(new CssClasses("ag-hidden"));
  const [getContractedCssClasses, setContractedCssClasses] = solidJs.createSignal(new CssClasses("ag-hidden"));
  const [getCheckboxCssClasses, setCheckboxCssClasses] = solidJs.createSignal(new CssClasses("ag-invisible"));
  props.ref(() => ({
    // force new instance when grid tries to refresh
    refresh() {
      return false;
    }
  }));
  solidJs.onMount(() => {
    const compProxy = {
      setInnerRenderer: (details, valueToDisplay) => {
        setInnerCompDetails(details);
        const escapedValue = agGridCommunity._.escapeString(valueToDisplay, true);
        setValue(escapedValue);
      },
      setChildCount: (count) => setChildCount(count),
      addOrRemoveCssClass: (name, on) => setCssClasses(getCssClasses().setClass(name, on)),
      setContractedDisplayed: (displayed) => setContractedCssClasses(getContractedCssClasses().setClass("ag-hidden", !displayed)),
      setExpandedDisplayed: (displayed) => setExpandedCssClasses(getExpandedCssClasses().setClass("ag-hidden", !displayed)),
      setCheckboxVisible: (visible) => setCheckboxCssClasses(getCheckboxCssClasses().setClass("ag-invisible", !visible))
    };
    const ctrl = context.createBean(new agGridCommunity.GroupCellRendererCtrl());
    ctrl.init(compProxy, eGui, eCheckboxRef, eExpandedRef, eContractedRef, GroupCellRenderer, props);
    eGui.setAttribute("role", ctrl.getCellAriaRole());
    role = ctrl.getCellAriaRole();
    return () => {
      context.destroyBean(ctrl);
    };
  });
  const getClassName = solidJs.createMemo(() => `ag-cell-wrapper ${getCssClasses().toString()}`);
  const getExpandedClassName = solidJs.createMemo(() => `ag-group-expanded ${getExpandedCssClasses().toString()}`);
  const getContractedClassName = solidJs.createMemo(() => `ag-group-contracted ${getContractedCssClasses().toString()}`);
  const getCheckboxClassName = solidJs.createMemo(() => `ag-group-checkbox ${getCheckboxCssClasses().toString()}`);
  const isShowUserComp = () => getInnerCompDetails() != null;
  const isShowValue = () => getInnerCompDetails() == null && getValue() != null;
  return web.ssrElement("span", web.mergeProps({
    get ["class"]() {
      return getClassName();
    }
  }, () => !props.colDef ? {
    role
  } : {}), [web.ssr(_tmpl$2, web.ssrAttribute("class", web.escape(getExpandedClassName(), true), false)), web.ssr(_tmpl$2, web.ssrAttribute("class", web.escape(getContractedClassName(), true), false)), web.ssr(_tmpl$2, web.ssrAttribute("class", web.escape(getCheckboxClassName(), true), false)), web.ssr(_tmpl$22, isShowUserComp() && web.escape(web.createComponent(userComp_default, {
    get compDetails() {
      return getInnerCompDetails();
    }
  })), isShowValue() && web.escape(getValue())), web.ssr(_tmpl$3, web.escape(getChildCount()))], false);
};
var groupCellRenderer_default = GroupCellRenderer;

// src/grid/core/solidFrameworkOverrides.tsx
var SolidFrameworkOverrides = class extends agGridCommunity.VanillaFrameworkOverrides {
  constructor() {
    super("solid");
  }
  frameworkComponents = {
    agGroupCellRenderer: groupCellRenderer_default,
    agGroupRowRenderer: groupCellRenderer_default,
    agDetailCellRenderer: detailCellRenderer_default
  };
  frameworkComponent(name) {
    return this.frameworkComponents[name];
  }
  isFrameworkComponent(comp) {
    if (!comp) {
      return false;
    }
    const prototype = comp.prototype;
    const isJsComp = prototype && "getGui" in prototype;
    return !isJsComp;
  }
};
var _tmpl$4 = ['<div class="ag-header-cell" style="', '"', ' role="columnheader"><div class="ag-header-cell-resize" role="presentation"></div><div class="ag-header-cell-comp-wrapper" role="presentation">', "</div></div>"];
var HeaderCellComp = (props) => {
  const {
    ctrl
  } = props;
  const [getWidth, setWidth] = solidJs.createSignal();
  const [getColId, setColId] = solidJs.createSignal(ctrl.getColId());
  const [getAriaSort, setAriaSort] = solidJs.createSignal();
  const [getUserCompDetails, setUserCompDetails] = solidJs.createSignal();
  let eGui;
  let eResize;
  let eHeaderCompWrapper;
  let userComp;
  const cssClassManager = new agGridCommunity.CssClassManager(() => eGui);
  solidJs.onMount(() => {
    const compProxy = {
      setWidth: (width) => setWidth(width),
      addOrRemoveCssClass: (name, on) => cssClassManager.addOrRemoveCssClass(name, on),
      setAriaSort: (sort) => setAriaSort(sort),
      setUserCompDetails: (compDetails) => setUserCompDetails(compDetails),
      getUserCompInstance: () => userComp
    };
    ctrl.setComp(compProxy, eGui, eResize, eHeaderCompWrapper);
    const selectAllGui = ctrl.getSelectAllGui();
    eResize.insertAdjacentElement("afterend", selectAllGui);
    ctrl.setDragSource(eGui);
  });
  const style = solidJs.createMemo(() => ({
    width: getWidth()
  }));
  return web.ssr(_tmpl$4, web.ssrStyle(style()), web.ssrAttribute("col-id", web.escape(getColId(), true), false) + web.ssrAttribute("aria-sort", web.escape(getAriaSort(), true), false), getUserCompDetails() && web.escape(web.createComponent(userComp_default, {
    get compDetails() {
      return getUserCompDetails();
    }
  })));
};
var headerCellComp_default = HeaderCellComp;
var _tmpl$5 = ["<div", ' style="', '" role="gridcell"><div', ' role="presentation">', "</div><div", ' role="presentation"><button type="button" class="ag-button ag-floating-filter-button-button"', "></button></div></div>"];
var HeaderFilterCellComp = (props) => {
  const [getCssClasses, setCssClasses] = solidJs.createSignal(new CssClasses());
  const [getCssBodyClasses, setBodyCssClasses] = solidJs.createSignal(new CssClasses());
  const [getCssButtonWrapperClasses, setButtonWrapperCssClasses] = solidJs.createSignal(new CssClasses());
  const [getButtonWrapperAriaHidden, setButtonWrapperAriaHidden] = solidJs.createSignal("false");
  const [getWidth, setWidth] = solidJs.createSignal();
  const [getUserCompDetails, setUserCompDetails] = solidJs.createSignal();
  let eGui;
  let eFloatingFilterBody;
  let eButtonShowMainFilter;
  let userCompPromise;
  solidJs.onMount(() => {
    userCompPromise = new agGridCommunity.AgPromise((resolve) => {
    });
  });
  const {
    ctrl
  } = props;
  solidJs.onMount(() => {
    const compProxy = {
      addOrRemoveCssClass: (name, on) => setCssClasses((prev) => prev.setClass(name, on)),
      addOrRemoveBodyCssClass: (name, on) => setBodyCssClasses((prev) => prev.setClass(name, on)),
      setButtonWrapperDisplayed: (displayed) => {
        setButtonWrapperCssClasses((prev) => prev.setClass("ag-hidden", !displayed));
        setButtonWrapperAriaHidden(!displayed ? "true" : "false");
      },
      setWidth: (width) => setWidth(width),
      setCompDetails: (compDetails) => setUserCompDetails(compDetails),
      getFloatingFilterComp: () => userCompPromise,
      setMenuIcon: (eIcon) => eButtonShowMainFilter.appendChild(eIcon)
    };
    ctrl.setComp(compProxy, eGui, eButtonShowMainFilter, eFloatingFilterBody);
  });
  const getStyle = solidJs.createMemo(() => ({
    width: getWidth()
  }));
  const getCssClassesString = solidJs.createMemo(() => "ag-header-cell ag-floating-filter " + getCssClasses().toString());
  const getBodyCssClassesString = solidJs.createMemo(() => getCssBodyClasses().toString());
  const getButtonWrapperCssClassesString = solidJs.createMemo(() => "ag-floating-filter-button " + getCssButtonWrapperClasses().toString());
  return web.ssr(_tmpl$5, web.ssrAttribute("class", web.escape(getCssClassesString(), true), false), web.ssrStyle(getStyle()), web.ssrAttribute("class", web.escape(getBodyCssClassesString(), true), false), getUserCompDetails() && web.escape(web.createComponent(userComp_default, {
    get compDetails() {
      return getUserCompDetails();
    }
  })), web.ssrAttribute("aria-hidden", web.escape(getButtonWrapperAriaHidden(), true), false) + web.ssrAttribute("class", web.escape(getButtonWrapperCssClassesString(), true), false), web.ssrAttribute("tabindex", -1, false));
};
var headerFilterCellComp_default = HeaderFilterCellComp;
var _tmpl$6 = ["<div", ' style="', '"', ' role="columnheader"', ">", "<div", "></div></div>"];
var HeaderGroupCellComp = (props) => {
  const {
    ctrl
  } = props;
  const [getCssClasses, setCssClasses] = solidJs.createSignal(new CssClasses());
  const [getCssResizableClasses, setResizableCssClasses] = solidJs.createSignal(new CssClasses());
  const [getResizableAriaHidden, setResizableAriaHidden] = solidJs.createSignal("false");
  const [getWidth, setWidth] = solidJs.createSignal();
  const [getColId, setColId] = solidJs.createSignal(ctrl.getColId());
  const [getAriaExpanded, setAriaExpanded] = solidJs.createSignal();
  const [getUserCompDetails, setUserCompDetails] = solidJs.createSignal();
  let eGui;
  let eResize;
  solidJs.onMount(() => {
    const compProxy = {
      setWidth: (width) => setWidth(width),
      addOrRemoveCssClass: (name, on) => setCssClasses(getCssClasses().setClass(name, on)),
      setUserCompDetails: (compDetails) => setUserCompDetails(compDetails),
      setResizableDisplayed: (displayed) => {
        setResizableCssClasses((prev) => prev.setClass("ag-hidden", !displayed));
        setResizableAriaHidden(!displayed ? "true" : "false");
      },
      setAriaExpanded: (expanded) => setAriaExpanded(expanded)
    };
    ctrl.setComp(compProxy, eGui, eResize);
  });
  solidJs.createEffect(() => {
    const userCompDetails = getUserCompDetails();
    if (userCompDetails == null) {
      return;
    }
    ctrl.setDragSource(eGui);
  });
  const style = solidJs.createMemo(() => ({
    width: getWidth()
  }));
  const getClassName = solidJs.createMemo(() => "ag-header-group-cell " + getCssClasses().toString());
  const getResizableClassName = solidJs.createMemo(() => "ag-header-cell-resize " + getCssResizableClasses().toString());
  return web.ssr(_tmpl$6, web.ssrAttribute("class", web.escape(getClassName(), true), false), web.ssrStyle(style()), web.ssrAttribute("col-id", web.escape(getColId(), true), false), web.ssrAttribute("aria-expanded", web.escape(getAriaExpanded(), true), false), getUserCompDetails() && web.escape(web.createComponent(userComp_default, {
    get compDetails() {
      return getUserCompDetails();
    }
  })), web.ssrAttribute("aria-hidden", web.escape(getResizableAriaHidden(), true), false) + web.ssrAttribute("class", web.escape(getResizableClassName(), true), false));
};
var headerGroupCellComp_default = HeaderGroupCellComp;

// src/grid/header/headerRowComp.tsx
var _tmpl$7 = ["<div", ' role="row" style="', '"', ">", "</div>"];
var HeaderRowComp = (props) => {
  const {
    gridOptionsService
  } = solidJs.useContext(BeansContext);
  const {
    ctrl
  } = props;
  const [getHeight, setHeight] = solidJs.createSignal();
  const [getTop, setTop] = solidJs.createSignal();
  const [getWidth, setWidth] = solidJs.createSignal();
  const [getAriaRowIndex, setAriaRowIndex] = solidJs.createSignal(ctrl.getAriaRowIndex());
  const [getCellCtrls, setCellCtrls] = solidJs.createSignal([]);
  const setCellCtrlsMaintainOrder = (next) => {
    const prev = getCellCtrls();
    const isEnsureDomOrder = gridOptionsService.get("ensureDomOrder");
    const isPrintLayout = gridOptionsService.isDomLayout("print");
    if (isEnsureDomOrder || isPrintLayout) {
      return next;
    }
    const prevMap = agGridCommunity._.mapById(prev, (c) => c.getInstanceId());
    const nextMap = agGridCommunity._.mapById(next, (c) => c.getInstanceId());
    const oldCtrlsWeAreKeeping = prev.filter((c) => nextMap.has(c.getInstanceId()));
    const newCtrls = next.filter((c) => !prevMap.has(c.getInstanceId()));
    const nextOrderMaintained = [...oldCtrlsWeAreKeeping, ...newCtrls];
    setCellCtrls(nextOrderMaintained);
  };
  solidJs.onMount(() => {
    const compProxy = {
      setHeight: (height) => setHeight(height),
      setTop: (top) => setTop(top),
      setHeaderCtrls: (ctrls) => setCellCtrlsMaintainOrder(ctrls),
      setWidth: (width) => setWidth(width)
    };
    ctrl.setComp(compProxy);
  });
  const style = solidJs.createMemo(() => ({
    height: getHeight(),
    top: getTop(),
    width: getWidth()
  }));
  const cssClasses = ctrl.getHeaderRowClass();
  const createCellJsx = (cellCtrl) => {
    switch (ctrl.getType()) {
      case agGridCommunity.HeaderRowType.COLUMN_GROUP:
        return web.createComponent(headerGroupCellComp_default, {
          ctrl: cellCtrl
        });
      case agGridCommunity.HeaderRowType.FLOATING_FILTER:
        return web.createComponent(headerFilterCellComp_default, {
          ctrl: cellCtrl
        });
      default:
        return web.createComponent(headerCellComp_default, {
          ctrl: cellCtrl
        });
    }
  };
  return web.ssr(_tmpl$7, web.ssrAttribute("class", web.escape(cssClasses, true), false), web.ssrStyle(style()), web.ssrAttribute("aria-rowindex", web.escape(getAriaRowIndex(), true), false), web.escape(web.createComponent(solidJs.For, {
    get each() {
      return getCellCtrls();
    },
    children: (cellCtrl, i) => createCellJsx(cellCtrl)
  })));
};
var headerRowComp_default = HeaderRowComp;

// src/grid/header/headerRowContainerComp.tsx
var _tmpl$8 = ["<div", ' role="rowgroup" style="', '">', "</div>"];
var _tmpl$23 = ["<div", ' role="presentation"><div class="ag-header-container" role="rowgroup" style="', '">', "</div></div>"];
var HeaderRowContainerComp = (props) => {
  const [getCssClasses, setCssClasses] = solidJs.createSignal(new CssClasses());
  const [getAriaHidden, setAriaHidden] = solidJs.createSignal(false);
  const [getCenterContainerWidth, setCenterContainerWidth] = solidJs.createSignal();
  const [getPinnedContainerWidth, setPinnedContainerWidth] = solidJs.createSignal();
  const [getHeaderRowCtrls, setHeaderRowCtrls] = solidJs.createSignal([]);
  const {
    context
  } = solidJs.useContext(BeansContext);
  let eGui;
  const pinnedLeft = props.pinned === "left";
  const pinnedRight = props.pinned === "right";
  const centre = !pinnedLeft && !pinnedRight;
  const destroyFuncs = [];
  solidJs.onCleanup(() => {
    destroyFuncs.forEach((f) => f());
    destroyFuncs.length = 0;
  });
  solidJs.onMount(() => {
    const compProxy = {
      setDisplayed: (displayed) => {
        setCssClasses(getCssClasses().setClass("ag-hidden", !displayed));
        setAriaHidden(!displayed);
      },
      setCtrls: (ctrls) => setHeaderRowCtrls(ctrls),
      // centre only
      setCenterWidth: (width) => setCenterContainerWidth(width),
      setViewportScrollLeft: (left) => eGui.scrollLeft = left,
      // pinned only
      setPinnedContainerWidth: (width) => setPinnedContainerWidth(width)
    };
    const ctrl = context.createBean(new agGridCommunity.HeaderRowContainerCtrl(props.pinned));
    ctrl.setComp(compProxy, eGui);
    destroyFuncs.push(() => context.destroyBean(ctrl));
  });
  const getClassName = solidJs.createMemo(() => getCssClasses().toString());
  const insertRowsJsx = () => web.createComponent(solidJs.For, {
    get each() {
      return getHeaderRowCtrls();
    },
    children: (ctrl) => web.createComponent(headerRowComp_default, {
      ctrl
    })
  });
  const eCenterContainerStyle = solidJs.createMemo(() => ({
    width: getCenterContainerWidth()
  }));
  const ePinnedStyle = solidJs.createMemo(() => ({
    width: getPinnedContainerWidth(),
    "min-width": getPinnedContainerWidth(),
    "max-width": getPinnedContainerWidth()
  }));
  return [pinnedLeft && web.ssr(_tmpl$8, web.ssrAttribute("class", "ag-pinned-left-header " + web.escape(getClassName(), true), false) + web.ssrAttribute("aria-hidden", web.escape(getAriaHidden(), true), false), web.ssrStyle(ePinnedStyle()), web.escape(insertRowsJsx())), pinnedRight && web.ssr(_tmpl$8, web.ssrAttribute("class", "ag-pinned-right-header " + web.escape(getClassName(), true), false) + web.ssrAttribute("aria-hidden", web.escape(getAriaHidden(), true), false), web.ssrStyle(ePinnedStyle()), web.escape(insertRowsJsx())), centre && web.ssr(_tmpl$23, web.ssrAttribute("class", "ag-header-viewport " + web.escape(getClassName(), true), false), web.ssrStyle(eCenterContainerStyle()), web.escape(insertRowsJsx()))];
};
var headerRowContainerComp_default = HeaderRowContainerComp;

// src/grid/header/gridHeaderComp.tsx
var _tmpl$9 = ["<div", ' style="', '" role="presentation">', "", "", "</div>"];
var GridHeaderComp = () => {
  const [getCssClasses, setCssClasses] = solidJs.createSignal(new CssClasses());
  const [getHeight, setHeight] = solidJs.createSignal();
  const {
    context
  } = solidJs.useContext(BeansContext);
  let eGui;
  const destroyFuncs = [];
  solidJs.onCleanup(() => {
    destroyFuncs.forEach((f) => f());
    destroyFuncs.length = 0;
  });
  solidJs.onMount(() => {
    const compProxy = {
      addOrRemoveCssClass: (name, on) => setCssClasses(getCssClasses().setClass(name, on)),
      setHeightAndMinHeight: (height) => setHeight(height)
    };
    const ctrl = context.createBean(new agGridCommunity.GridHeaderCtrl());
    ctrl.setComp(compProxy, eGui, eGui);
    destroyFuncs.push(() => context.destroyBean(ctrl));
  });
  const className = solidJs.createMemo(() => {
    let res = getCssClasses().toString();
    return "ag-header " + res;
  });
  const style = solidJs.createMemo(() => ({
    height: getHeight(),
    "min-height": getHeight()
  }));
  return web.ssr(_tmpl$9, web.ssrAttribute("class", web.escape(className(), true), false), web.ssrStyle(style()), web.escape(web.createComponent(headerRowContainerComp_default, {
    pinned: "left"
  })), web.escape(web.createComponent(headerRowContainerComp_default, {
    pinned: null
  })), web.escape(web.createComponent(headerRowContainerComp_default, {
    pinned: "right"
  })));
};
var gridHeaderComp_default = GridHeaderComp;
var PopupEditorComp = (props) => {
  const {
    context,
    popupService,
    localeService,
    gridOptionsService
  } = solidJs.useContext(BeansContext);
  const {
    editDetails,
    cellCtrl,
    eParentCell
  } = props;
  const {
    compDetails
  } = editDetails;
  const useModelPopup = gridOptionsService.get("stopEditingWhenCellsLoseFocus");
  const wrapper = context.createBean(new agGridCommunity.PopupEditorWrapper(compDetails.params));
  const ePopupGui = wrapper.getGui();
  const positionParams = {
    column: cellCtrl.getColumn(),
    rowNode: cellCtrl.getRowNode(),
    type: "popupCellEditor",
    eventSource: eParentCell,
    ePopup: ePopupGui,
    position: editDetails.popupPosition,
    keepWithinBounds: true
  };
  const positionCallback = popupService.positionPopupByComponent.bind(popupService, positionParams);
  const translate = localeService.getLocaleTextFunc();
  const addPopupRes = popupService.addPopup({
    modal: useModelPopup,
    eChild: ePopupGui,
    closeOnEsc: true,
    closedCallback: () => {
      cellCtrl.onPopupEditorClosed();
    },
    anchorToElement: eParentCell,
    positionCallback,
    ariaLabel: translate("ariaLabelCellEditor", "Cell Editor")
  });
  const hideEditorPopup = addPopupRes ? addPopupRes.hideFunc : void 0;
  solidJs.onCleanup(() => {
    if (hideEditorPopup != null) {
      hideEditorPopup();
    }
    context.destroyBean(wrapper);
  });
  return web.createComponent(web.Portal, {
    mount: ePopupGui,
    get children() {
      return props.children;
    }
  });
};
var popupEditorComp_default = PopupEditorComp;

// src/grid/cells/showEditDetails.tsx
var ShowEditDetails = (props) => {
  const getCompDetails = solidJs.createMemo(() => props.editDetails.compDetails);
  const compDetails = props.editDetails.compDetails;
  if (!compDetails) {
    return [];
  }
  const inPopup = props.editDetails.popup;
  const eGui = props.eGuiFn();
  return [inPopup && web.createComponent(popupEditorComp_default, {
    get cellCtrl() {
      return props.cellCtrl;
    },
    eParentCell: eGui,
    get editDetails() {
      return props.editDetails;
    },
    get children() {
      return web.createComponent(userComp_default, {
        get compDetails() {
          return getCompDetails();
        }
      });
    }
  }), !inPopup && web.createComponent(userComp_default, {
    get compDetails() {
      return getCompDetails();
    }
  })];
};
var showEditDetails_default = ShowEditDetails;
var _tmpl$10 = ['<span role="presentation" id="', '" class="ag-cell-value">', "</span>"];
var ToolsComp = (props) => {
  const {
    context
  } = solidJs.useContext(BeansContext);
  const CompWrapper = (innerProps) => {
    const comp = innerProps.createFn();
    if (!comp) {
      return [];
    }
    solidJs.onCleanup(() => context.destroyBean(comp));
    return comp.getGui();
  };
  return [props.includeSelection && web.createComponent(CompWrapper, {
    createFn: () => {
      const checkboxSelectionComp = props.cellCtrl.createSelectionCheckbox();
      props.setSelectionCheckboxId(checkboxSelectionComp.getCheckboxId());
      return checkboxSelectionComp;
    }
  }), props.includeDndSource && web.createComponent(CompWrapper, {
    createFn: () => props.cellCtrl.createDndSource()
  }), props.includeRowDrag && web.createComponent(CompWrapper, {
    createFn: () => props.cellCtrl.createRowDragComp()
  })];
};
var ShowRenderDetails = (props) => {
  const getCompDetails = solidJs.createMemo(() => props.showDetails.compDetails);
  const isNoCompDetails = solidJs.createMemo(() => props.showDetails.compDetails == null);
  const valueForNoCellRenderer = () => {
    const value = props.showDetails.value;
    return value && value.toString ? value.toString() : value;
  };
  const bodyJsxFunc = () => [isNoCompDetails() && valueForNoCellRenderer(), getCompDetails() && web.createComponent(userComp_default, {
    get compDetails() {
      return getCompDetails();
    }
  })];
  return [props.showTools && web.createComponent(ToolsComp, {
    get includeDndSource() {
      return props.includeDndSource;
    },
    get includeRowDrag() {
      return props.includeRowDrag;
    },
    get includeSelection() {
      return props.includeSelection;
    },
    get setSelectionCheckboxId() {
      return props.setSelectionCheckboxId;
    },
    get cellCtrl() {
      return props.cellCtrl;
    }
  }), props.showCellWrapper ? web.ssr(_tmpl$10, `cell-${web.escape(props.cellInstanceId, true)}`, web.escape(bodyJsxFunc())) : bodyJsxFunc()];
};
var showRenderDetails_default = ShowRenderDetails;

// src/grid/cells/cellComp.tsx
var _tmpl$11 = ['<div style="', '"', "> ", "</div>"];
var _tmpl$24 = ['<div class="ag-cell-wrapper" role="presentation">', "</div>"];
var checkCellEditorDeprecations = (popup, cellEditor, cellCtrl) => {
  const col = cellCtrl.getColumn();
  if (!popup && cellEditor.isPopup && cellEditor.isPopup()) {
    `AG Grid: Found an issue in column ${col.getColId()}. If using SolidJS, specify an editor is a popup using colDef.cellEditorPopup=true. AG Grid SolidJS cannot depend on the editor component specifying if it's in a popup (via the isPopup() method on the editor), as SolidJS needs to know this information BEFORE the component is created.`;
    agGridCommunity._.doOnce(() => void 0, "jsEditorComp-isPopup-" + cellCtrl.getColumn().getColId());
  }
  if (popup && cellEditor.getPopupPosition && cellEditor.getPopupPosition() != null) {
    `AG Grid: Found an issue in column ${col.getColId()}. If using SolidJS, specify an editor popup position using colDef.cellEditorPopupPosition=true. AG Grid SolidJS cannot depend on the editor component specifying it's position (via the getPopupPosition() method on the editor), as SolidJS needs to know this information BEFORE the component is created.`;
    agGridCommunity._.doOnce(() => void 0, "jsEditorComp-getPopupPosition-" + cellCtrl.getColumn().getColId());
  }
};
var CellComp = (props) => {
  const {
    cellCtrl,
    printLayout,
    editingRow
  } = props;
  const [renderDetails, setRenderDetails] = solidJs.createSignal();
  const [editDetails, setEditDetails] = solidJs.createSignal();
  let renderCompVersion = 0;
  const [renderCompVersionList, setRenderCompVersionList] = solidJs.createSignal([renderCompVersion]);
  const [userStyles, setUserStyles] = solidJs.createSignal();
  const [tabIndex, setTabIndex] = solidJs.createSignal(cellCtrl.getTabIndex());
  const [colId, setColId] = solidJs.createSignal(cellCtrl.getColumnIdSanitised());
  const [selectionCheckboxId, setSelectionCheckboxId] = solidJs.createSignal();
  const [includeSelection, setIncludeSelection] = solidJs.createSignal(false);
  const [includeRowDrag, setIncludeRowDrag] = solidJs.createSignal(false);
  const [includeDndSource, setIncludeDndSource] = solidJs.createSignal(false);
  const forceWrapper = cellCtrl.isForceWrapper();
  let eCellWrapper;
  let eCellValue;
  const setECellValue = (val) => {
    eCellValue = val;
  };
  let eGui;
  let cellEditor = null;
  const setEditorRef = (popup, ref) => {
    cellEditor = ref;
    if (!cellEditor) {
      return;
    }
    checkCellEditorDeprecations(popup, cellEditor, cellCtrl);
    const editingCancelledByUserComp = cellEditor.isCancelBeforeStart && cellEditor.isCancelBeforeStart();
    if (editingCancelledByUserComp) {
      setTimeout(() => {
        cellCtrl.stopEditing(true);
        cellCtrl.focusCell(true);
      });
    }
    const refAny = ref;
    if (refAny.afterGuiAttached) {
      setTimeout(() => refAny.afterGuiAttached(), 0);
    }
  };
  const setPopupEditorRef = (ref) => setEditorRef(true, ref);
  const setInlineEditorRef = (ref) => setEditorRef(false, ref);
  const cssClassManager = new agGridCommunity.CssClassManager(() => eGui);
  const showTools = solidJs.createMemo(() => renderDetails() != null && (includeSelection() || includeDndSource() || includeRowDrag()));
  const showCellWrapper = solidJs.createMemo(() => forceWrapper || showTools());
  const cellInstanceId = cellCtrl.getInstanceId();
  solidJs.onMount(() => {
    if (!cellCtrl) {
      return;
    }
    const compProxy = {
      addOrRemoveCssClass: (name, on) => cssClassManager.addOrRemoveCssClass(name, on),
      setUserStyles: (styles) => setUserStyles(styles),
      getFocusableElement: () => eGui,
      setIncludeSelection: (include) => setIncludeSelection(include),
      setIncludeRowDrag: (include) => setIncludeRowDrag(include),
      setIncludeDndSource: (include) => setIncludeDndSource(include),
      getCellEditor: () => cellEditor,
      getCellRenderer: () => null,
      getParentOfValue: () => eCellValue ? eCellValue : eGui,
      setRenderDetails: (compDetails, value, force) => {
        setRenderDetails({
          value,
          compDetails,
          force
        });
      },
      setEditDetails: (compDetails, popup, popupPosition) => {
        if (compDetails) {
          setEditDetails({
            compDetails,
            popup,
            popupPosition
          });
          if (!popup) {
            setRenderDetails(void 0);
          }
        } else {
          setEditDetails(void 0);
        }
      }
    };
    cellCtrl.setComp(compProxy, eGui, eCellWrapper, printLayout, editingRow);
  });
  solidJs.createEffect(() => {
    const isEditing = !!editDetails();
    const isPopup = isEditing && !!editDetails()?.popup;
    cssClassManager.addOrRemoveCssClass("ag-cell-value", !showCellWrapper());
    cssClassManager.addOrRemoveCssClass("ag-cell-inline-editing", isEditing && !isPopup);
    cssClassManager.addOrRemoveCssClass("ag-cell-popup-editing", isEditing && isPopup);
    cssClassManager.addOrRemoveCssClass("ag-cell-not-inline-editing", !isEditing || isPopup);
    cellCtrl.getRowCtrl()?.setInlineEditingCss(isEditing);
  });
  let readyForRefresh = false;
  solidJs.createEffect(() => {
    const details = renderDetails();
    const isJsCellRenderer = details != null && details.compDetails != null && !details.compDetails.componentFromFramework;
    if (!isJsCellRenderer) {
      readyForRefresh = false;
      return;
    }
    if (!readyForRefresh) {
      readyForRefresh = true;
      return;
    }
    {
      return;
    }
  });
  const eGuiFn = () => eGui;
  const bodyJsxFunc = () => [web.createComponent(solidJs.For, {
    get each() {
      return renderCompVersionList();
    },
    children: () => renderDetails() && web.createComponent(showRenderDetails_default, {
      get showDetails() {
        return renderDetails();
      },
      cellInstanceId,
      get showCellWrapper() {
        return showCellWrapper();
      },
      cellCtrl,
      get includeDndSource() {
        return includeDndSource();
      },
      get includeRowDrag() {
        return includeRowDrag();
      },
      get includeSelection() {
        return includeSelection();
      },
      setSelectionCheckboxId,
      get showTools() {
        return showTools();
      },
      setECellValue
    })
  }), editDetails() && web.createComponent(showEditDetails_default, {
    get editDetails() {
      return editDetails();
    },
    cellCtrl,
    eGuiFn,
    setInlineRef: setInlineEditorRef,
    setPopupRef: setPopupEditorRef
  })];
  return web.ssr(_tmpl$11, web.ssrStyle(userStyles()), web.ssrAttribute("tabindex", web.escape(tabIndex(), true), false) + web.ssrAttribute("role", web.escape(cellCtrl.getCellAriaRole(), true), false) + web.ssrAttribute("col-id", web.escape(colId(), true), false), showCellWrapper() ? web.ssr(_tmpl$24, web.escape(bodyJsxFunc())) : web.escape(bodyJsxFunc()));
};
var cellComp_default = CellComp;

// src/grid/rows/rowComp.tsx
var _tmpl$12 = ['<div role="row" style="', '"', ">", "", "</div>"];
var maintainOrderOnColumns = (prev, next, domOrder) => {
  if (domOrder) {
    const res2 = {
      list: next,
      instanceIdMap: /* @__PURE__ */ new Map()
    };
    next.forEach((c) => res2.instanceIdMap.set(c.getInstanceId(), c));
    return res2;
  }
  const oldCellCtrls = [];
  const newCellCtrls = [];
  const newInstanceIdMap = /* @__PURE__ */ new Map();
  const tempMap = /* @__PURE__ */ new Map();
  next.forEach((c) => tempMap.set(c.getInstanceId(), c));
  prev.list.forEach((c) => {
    const instanceId = c.getInstanceId();
    if (tempMap.has(instanceId)) {
      oldCellCtrls.push(c);
      newInstanceIdMap.set(instanceId, c);
    }
  });
  next.forEach((c) => {
    const instanceId = c.getInstanceId();
    if (!prev.instanceIdMap.has(instanceId)) {
      newCellCtrls.push(c);
      newInstanceIdMap.set(instanceId, c);
    }
  });
  const res = {
    list: [...oldCellCtrls, ...newCellCtrls],
    instanceIdMap: newInstanceIdMap
  };
  return res;
};
var RowComp = (params) => {
  const {
    rowCtrl,
    containerType
  } = params;
  const [getRowIndex, setRowIndex] = solidJs.createSignal();
  const [getRowId, setRowId] = solidJs.createSignal();
  const [getRowBusinessKey, setRowBusinessKey] = solidJs.createSignal();
  const [getTabIndex, setTabIndex] = solidJs.createSignal(rowCtrl.getTabIndex());
  const [getUserStyles, setUserStyles] = solidJs.createSignal();
  const [getCellCtrls, setCellCtrls] = solidJs.createSignal({
    list: [],
    instanceIdMap: /* @__PURE__ */ new Map()
  });
  const [getFullWidthCompDetails, setFullWidthCompDetails] = solidJs.createSignal();
  const [getDomOrder, setDomOrder] = solidJs.createSignal(false);
  const [getTop, setTop] = solidJs.createSignal(rowCtrl.getInitialRowTop(containerType));
  const [getTransform, setTransform] = solidJs.createSignal(rowCtrl.getInitialTransform(containerType));
  let eGui;
  let fullWidthCompRef;
  solidJs.createEffect(() => {
    const compDetails = getFullWidthCompDetails();
    if (!compDetails) {
      return;
    }
    let tryCount = 0;
    const trySetup = () => {
      eGui.firstChild;
      if (tryCount >= 10) {
        return;
      }
      tryCount++;
      setTimeout(trySetup, 0);
    };
    trySetup();
  });
  solidJs.onMount(() => {
    if (!rowCtrl.isAlive()) {
      return;
    }
    const cssClassManager = new agGridCommunity.CssClassManager(() => eGui);
    const compProxy = {
      // the rowTop is managed by state, instead of direct style manipulation by rowCtrl (like all the other styles)
      // as we need to have an initial value when it's placed into he DOM for the first time, for animation to work.
      setTop: (value) => setTop(value),
      setTransform: (value) => setTransform(value),
      // i found using React for managing classes at the row level was to slow, as modifying classes caused a lot of
      // React code to execute, so avoiding React for managing CSS Classes made the grid go much faster.
      addOrRemoveCssClass: (name, on) => cssClassManager.addOrRemoveCssClass(name, on),
      setDomOrder: (domOrder) => setDomOrder(domOrder),
      setRowIndex: (value) => setRowIndex(value),
      setRowId: (value) => setRowId(value),
      setRowBusinessKey: (value) => setRowBusinessKey(value),
      setUserStyles: (styles) => setUserStyles(styles),
      // if we don't maintain the order, then cols will be ripped out and into the dom
      // when cols reordered, which would stop the CSS transitions from working
      setCellCtrls: (next) => setCellCtrls(maintainOrderOnColumns(getCellCtrls(), next, getDomOrder())),
      showFullWidth: (compDetails) => setFullWidthCompDetails(compDetails),
      getFullWidthCellRenderer: () => fullWidthCompRef,
      refreshFullWidth: (getUpdatedParams) => {
        {
          return false;
        }
      }
    };
    rowCtrl.setComp(compProxy, eGui, containerType);
    solidJs.onCleanup(() => rowCtrl.unsetComp(containerType));
  });
  const getRowStyles = solidJs.createMemo(() => {
    const res = {
      top: getTop(),
      transform: getTransform()
    };
    Object.assign(res, getUserStyles());
    return res;
  });
  const isShowCells = solidJs.createMemo(() => getCellCtrls() != null);
  const isShowFullWidth = solidJs.createMemo(() => getFullWidthCompDetails() != null);
  const showCellsJsx = () => web.createComponent(solidJs.For, {
    get each() {
      return getCellCtrls().list;
    },
    children: (cellCtrl) => web.createComponent(cellComp_default, {
      cellCtrl,
      get editingRow() {
        return rowCtrl.isEditing();
      },
      get printLayout() {
        return rowCtrl.isPrintLayout();
      }
    })
  });
  const showFullWidthJsx = () => web.createComponent(userComp_default, {
    get compDetails() {
      return getFullWidthCompDetails();
    }
  });
  return web.ssr(_tmpl$12, web.ssrStyle(getRowStyles()), web.ssrAttribute("row-index", web.escape(getRowIndex(), true), false) + web.ssrAttribute("row-id", web.escape(getRowId(), true), false) + web.ssrAttribute("row-business-key", web.escape(getRowBusinessKey(), true), false) + web.ssrAttribute("tabindex", web.escape(getTabIndex(), true), false), isShowFullWidth() && web.escape(showFullWidthJsx()), isShowCells() && web.escape(showCellsJsx()));
};
var rowComp_default = RowComp;

// src/grid/rows/rowContainerComp.tsx
var _tmpl$13 = ["<div", ' role="rowgroup">', "</div>"];
var _tmpl$25 = ["<div", ' role="presentation" style="', '">', "</div>"];
var RowContainerComp = (props) => {
  const {
    context
  } = solidJs.useContext(BeansContext);
  const [viewportHeight, setViewportHeight] = solidJs.createSignal("");
  const [rowCtrlsOrdered, setRowCtrlsOrdered] = solidJs.createSignal([]);
  const [rowCtrls, setRowCtrls] = solidJs.createSignal([]);
  const [domOrder, setDomOrder] = solidJs.createSignal(false);
  const {
    name
  } = props;
  const containerType = solidJs.createMemo(() => agGridCommunity.getRowContainerTypeForName(name));
  let eViewport;
  let eContainer;
  const cssClasses = solidJs.createMemo(() => agGridCommunity.RowContainerCtrl.getRowContainerCssClasses(name));
  const viewportClasses = solidJs.createMemo(() => classesList(cssClasses().viewport));
  const containerClasses = solidJs.createMemo(() => classesList(cssClasses().container));
  const centerTemplate = name === agGridCommunity.RowContainerName.CENTER || name === agGridCommunity.RowContainerName.TOP_CENTER || name === agGridCommunity.RowContainerName.BOTTOM_CENTER || name === agGridCommunity.RowContainerName.STICKY_TOP_CENTER;
  let rowCtrlsOrderedCopy = [];
  solidJs.createEffect(() => {
    if (domOrder()) {
      setRowCtrlsOrdered(rowCtrls());
      return;
    }
    const prev = rowCtrlsOrderedCopy;
    const oldRows = prev.filter((r) => rowCtrls().indexOf(r) >= 0);
    const newRows = rowCtrls().filter((r) => oldRows.indexOf(r) < 0);
    const next = [...oldRows, ...newRows];
    setRowCtrlsOrdered(next);
    rowCtrlsOrderedCopy = next;
  });
  solidJs.onMount(() => {
    const compProxy = {
      setViewportHeight,
      setRowCtrls: ({
        rowCtrls: rowCtrls2
      }) => setRowCtrls(rowCtrls2),
      setDomOrder: (domOrder2) => setDomOrder(domOrder2),
      setContainerWidth: (width) => {
      }
    };
    const ctrl = context.createBean(new agGridCommunity.RowContainerCtrl(name));
    solidJs.onCleanup(() => context.destroyBean(ctrl));
    ctrl.setComp(compProxy, eContainer, eViewport);
  });
  const viewportStyle = solidJs.createMemo(() => ({
    height: viewportHeight()
  }));
  const buildContainer = () => web.ssr(_tmpl$13, web.ssrAttribute("class", web.escape(containerClasses(), true), false), web.escape(web.createComponent(solidJs.For, {
    get each() {
      return rowCtrlsOrdered();
    },
    children: (rowCtrl, i) => web.createComponent(rowComp_default, {
      rowCtrl,
      get containerType() {
        return containerType();
      }
    })
  })));
  return centerTemplate ? web.ssr(_tmpl$25, web.ssrAttribute("class", web.escape(viewportClasses(), true), false), web.ssrStyle(viewportStyle()), web.escape(buildContainer())) : buildContainer();
};
var rowContainerComp_default = RowContainerComp;

// src/grid/gridBodyComp.tsx
var _tmpl$14 = ["<div", ' role="treegrid"', ">", "<div", ' role="presentation" style="', '">', "", "", "", "</div><div", ' role="presentation"><div', ' role="presentation" style="', '">', "", "", "", "</div></div><div", ' role="presentation" style="', '">', "", "", "", "</div><div", ' role="presentation" style="', '">', "", "", "", "</div></div>"];
var GridBodyComp = () => {
  const {
    context,
    agStackComponentsRegistry,
    resizeObserverService
  } = solidJs.useContext(BeansContext);
  const [getRowAnimationClass, setRowAnimationClass] = solidJs.createSignal("");
  const [getAriaColCount, setAriaColCount] = solidJs.createSignal(0);
  const [getAriaRowCount, setAriaRowCount] = solidJs.createSignal(0);
  const [getTopHeight, setTopHeight] = solidJs.createSignal(0);
  const [getBottomHeight, setBottomHeight] = solidJs.createSignal(0);
  const [getStickyTopHeight, setStickyTopHeight] = solidJs.createSignal("0px");
  const [getStickyTopTop, setStickyTopTop] = solidJs.createSignal("0px");
  const [getStickyTopWidth, setStickyTopWidth] = solidJs.createSignal("100%");
  const [getTopDisplay, setTopDisplay] = solidJs.createSignal("");
  const [getBottomDisplay, setBottomDisplay] = solidJs.createSignal("");
  const [getBodyViewportWidth, setBodyViewportWidth] = solidJs.createSignal("");
  const [getMovingCss, setMovingCss] = solidJs.createSignal(null);
  const [getForceVerticalScrollClass, setForceVerticalScrollClass] = solidJs.createSignal(null);
  const [getTopAndBottomOverflowY, setTopAndBottomOverflowY] = solidJs.createSignal(null);
  const [getCellSelectableCss, setCellSelectableCss] = solidJs.createSignal(null);
  const [getLayoutClass, setLayoutClass] = solidJs.createSignal("ag-layout-normal");
  let eRoot;
  let eTop;
  let eStickyTop;
  let eBody;
  let eBodyViewport;
  let eBottom;
  const destroyFuncs = [];
  solidJs.onCleanup(() => {
    destroyFuncs.forEach((f) => f());
    destroyFuncs.length = 0;
  });
  solidJs.onMount(() => {
    if (!context) {
      return;
    }
    const newComp = (tag) => {
      const CompClass = agStackComponentsRegistry.getComponentClass(tag);
      const comp = context.createBean(new CompClass());
      solidJs.onCleanup(() => context.destroyBean(comp));
      return comp;
    };
    eRoot.appendChild(newComp("AG-FAKE-HORIZONTAL-SCROLL").getGui());
    eRoot.appendChild(newComp("AG-OVERLAY-WRAPPER").getGui());
    eBody.appendChild(newComp("AG-FAKE-VERTICAL-SCROLL").getGui());
    const compProxy = {
      setRowAnimationCssOnBodyViewport: setRowAnimationClass,
      setColumnCount: setAriaColCount,
      setRowCount: setAriaRowCount,
      setTopHeight,
      setBottomHeight,
      setStickyTopHeight,
      setStickyTopTop,
      setStickyTopWidth,
      setTopDisplay,
      setBottomDisplay,
      setColumnMovingCss: setMovingCss,
      updateLayoutClasses: setLayoutClass,
      setAlwaysVerticalScrollClass: setForceVerticalScrollClass,
      setPinnedTopBottomOverflowY: setTopAndBottomOverflowY,
      setCellSelectableCss: (cssClass, flag) => setCellSelectableCss(flag ? cssClass : null),
      setBodyViewportWidth,
      registerBodyViewportResizeListener: (listener) => {
        const unsubscribeFromResize = resizeObserverService.observeResize(eBodyViewport, listener);
        destroyFuncs.push(() => unsubscribeFromResize());
      }
    };
    const ctrl = context.createBean(new agGridCommunity.GridBodyCtrl());
    solidJs.onCleanup(() => context.destroyBean(ctrl));
    setTimeout(() => ctrl.setComp(compProxy, eRoot, eBodyViewport, eTop, eBottom, eStickyTop), 0);
  });
  const getRootClasses = solidJs.createMemo(() => classesList("ag-root", "ag-unselectable", getMovingCss(), getLayoutClass()));
  const getBodyClasses = solidJs.createMemo(() => classesList("ag-body", getLayoutClass()));
  const getBodyViewportClasses = solidJs.createMemo(() => classesList("ag-body-viewport", getRowAnimationClass(), getLayoutClass(), getForceVerticalScrollClass(), getCellSelectableCss()));
  const getTopClasses = solidJs.createMemo(() => classesList("ag-floating-top", getCellSelectableCss()));
  const getStickyTopClasses = solidJs.createMemo(() => classesList("ag-sticky-top", getCellSelectableCss()));
  const getBottomClasses = solidJs.createMemo(() => classesList("ag-floating-bottom", getCellSelectableCss()));
  const getTopStyle = solidJs.createMemo(() => ({
    height: `${getTopHeight()}px`,
    "min-height": `${getTopHeight()}px`,
    display: getTopDisplay(),
    "overflow-y": getTopAndBottomOverflowY()
  }));
  const getStickyTopStyle = solidJs.createMemo(() => ({
    height: getStickyTopHeight(),
    top: getStickyTopTop(),
    width: getStickyTopWidth()
  }));
  const getBottomStyle = solidJs.createMemo(() => ({
    height: `${getBottomHeight()}px`,
    "min-height": `${getBottomHeight()}px`,
    display: getBottomDisplay(),
    "overflow-y": getTopAndBottomOverflowY()
  }));
  const getBodyViewportStyle = solidJs.createMemo(() => ({
    width: getBodyViewportWidth()
  }));
  return web.ssr(_tmpl$14, web.ssrAttribute("class", web.escape(getRootClasses(), true), false), web.ssrAttribute("aria-colcount", web.escape(getAriaColCount(), true), false) + web.ssrAttribute("aria-rowcount", web.escape(getAriaRowCount(), true), false), web.escape(web.createComponent(gridHeaderComp_default, {})), web.ssrAttribute("class", web.escape(getTopClasses(), true), false), web.ssrStyle(getTopStyle()), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.TOP_LEFT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.TOP_CENTER;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.TOP_RIGHT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.TOP_FULL_WIDTH;
    }
  })), web.ssrAttribute("class", web.escape(getBodyClasses(), true), false), web.ssrAttribute("class", web.escape(getBodyViewportClasses(), true), false), web.ssrStyle(getBodyViewportStyle()), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.LEFT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.CENTER;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.RIGHT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.FULL_WIDTH;
    }
  })), web.ssrAttribute("class", web.escape(getStickyTopClasses(), true), false), web.ssrStyle(getStickyTopStyle()), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.STICKY_TOP_LEFT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.STICKY_TOP_CENTER;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.STICKY_TOP_RIGHT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.STICKY_TOP_FULL_WIDTH;
    }
  })), web.ssrAttribute("class", web.escape(getBottomClasses(), true), false), web.ssrStyle(getBottomStyle()), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.BOTTOM_LEFT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.BOTTOM_CENTER;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.BOTTOM_RIGHT;
    }
  })), web.escape(web.createComponent(rowContainerComp_default, {
    get name() {
      return agGridCommunity.RowContainerName.BOTTOM_FULL_WIDTH;
    }
  })));
};
var gridBodyComp_default = GridBodyComp;
var _tmpl$15 = ['<div class="', '" role="presentation"', "></div>"];
var TabGuardComp = (props) => {
  const {
    children,
    eFocusableElement,
    onTabKeyDown,
    gridCtrl,
    forceFocusOutWhenTabGuardsAreEmpty
  } = props;
  const [tabIndex, setTabIndex] = solidJs.createSignal();
  let eTopGuard;
  let eBottomGuard;
  let ctrl;
  const {
    context
  } = solidJs.useContext(BeansContext);
  solidJs.onMount(() => {
    const compProxy = {
      setTabIndex: (value) => value == null ? setTabIndex(void 0) : setTabIndex(parseInt(value, 10))
    };
    ctrl = context.createBean(new agGridCommunity.TabGuardCtrl({
      comp: compProxy,
      eTopGuard,
      eBottomGuard,
      eFocusableElement,
      onTabKeyDown,
      forceFocusOutWhenTabGuardsAreEmpty,
      focusInnerElement: (fromBottom) => gridCtrl.focusInnerElement(fromBottom)
    }));
    props.ref({
      forceFocusOutOfContainer(up) {
        ctrl.forceFocusOutOfContainer(up);
      }
    });
  });
  solidJs.onCleanup(() => context.destroyBean(ctrl));
  return [web.ssr(_tmpl$15, `${web.escape(agGridCommunity.TabGuardClassNames.TAB_GUARD, true)} ${web.escape(agGridCommunity.TabGuardClassNames.TAB_GUARD_TOP, true)}`, web.ssrAttribute("tabindex", web.escape(tabIndex(), true), false)), children, web.ssr(_tmpl$15, `${web.escape(agGridCommunity.TabGuardClassNames.TAB_GUARD, true)} ${web.escape(agGridCommunity.TabGuardClassNames.TAB_GUARD_BOTTOM, true)}`, web.ssrAttribute("tabindex", web.escape(tabIndex(), true), false))];
};
var tabGuardComp_default = TabGuardComp;

// src/grid/gridComp.tsx
var _tmpl$16 = ["<div", ' style="', '"><div', ">", "</div></div>"];
var GridComp = (props) => {
  const [rtlClass, setRtlClass] = solidJs.createSignal("");
  const [keyboardFocusClass, setKeyboardFocusClass] = solidJs.createSignal("");
  const [layoutClass, setLayoutClass] = solidJs.createSignal("");
  const [cursor, setCursor] = solidJs.createSignal(null);
  const [userSelect, setUserSelect] = solidJs.createSignal(null);
  const [initialised, setInitialised] = solidJs.createSignal(false);
  const {
    context
  } = props;
  const beans = context.getBean("beans");
  let eGui;
  let gridCtrl;
  const onTabKeyDown = () => void 0;
  const destroyFuncs = [];
  solidJs.onCleanup(() => {
    destroyFuncs.forEach((f) => f());
    destroyFuncs.length = 0;
  });
  solidJs.onMount(() => {
    gridCtrl = context.createBean(new agGridCommunity.GridCtrl());
    destroyFuncs.push(() => context.destroyBean(gridCtrl));
    const compProxy = {
      destroyGridUi: () => {
      },
      // do nothing, as framework users destroy grid by removing the comp
      setRtlClass,
      forceFocusOutOfContainer: (up) => {
      },
      updateLayoutClasses: setLayoutClass,
      getFocusableContainers: () => {
        const els = [];
        const gridBodyCompEl = eGui.querySelector(".ag-root");
        const sideBarEl = eGui.querySelector(".ag-side-bar:not(.ag-hidden)");
        if (gridBodyCompEl) {
          els.push(gridBodyCompEl);
        }
        if (sideBarEl) {
          els.push(sideBarEl);
        }
        return els;
      },
      setCursor,
      setUserSelect
    };
    gridCtrl.setComp(compProxy, eGui, eGui);
    setInitialised(true);
  });
  const cssClasses = solidJs.createMemo(() => classesList("ag-root-wrapper", rtlClass(), keyboardFocusClass(), layoutClass(), props.class));
  const bodyCssClasses = solidJs.createMemo(() => classesList("ag-root-wrapper-body", "ag-focus-managed", layoutClass()));
  const topStyle = solidJs.createMemo(() => ({
    userSelect: userSelect != null ? userSelect() : "",
    WebkitUserSelect: userSelect != null ? userSelect() : "",
    cursor: cursor != null ? cursor() : ""
  }));
  return web.ssr(_tmpl$16, web.ssrAttribute("class", web.escape(cssClasses(), true), false), web.ssrStyle(topStyle()), web.ssrAttribute("class", web.escape(bodyCssClasses(), true), false), initialised() && web.escape(
    // we wait for initialised before rending the children, so GridComp has created and registered with it's
    // GridCtrl before we create the child GridBodyComp. Otherwise the GridBodyComp would initialise first,
    // before we have set the the Layout CSS classes, causing the GridBodyComp to render rows to a grid that
    // doesn't have it's height specified, which would result if all the rows getting rendered (and if many rows,
    // hangs the UI)
    web.createComponent(BeansContext.Provider, {
      value: beans,
      get children() {
        return web.createComponent(tabGuardComp_default, {
          eFocusableElement: eGui,
          onTabKeyDown,
          gridCtrl,
          forceFocusOutWhenTabGuardsAreEmpty: true,
          get children() {
            return web.createComponent(gridBodyComp_default, {});
          }
        });
      }
    })
  ));
};
var gridComp_default = GridComp;

// src/grid/agGridSolid.tsx
var _tmpl$17 = ['<div style="', '">', "", "</div>"];
var AgGridSolid = function(props) {
  let eGui;
  let api;
  const [context, setContext] = solidJs.createSignal();
  const [getPortals, setPortals] = solidJs.createSignal([]);
  const destroyFuncs = [];
  solidJs.onCleanup(() => {
    destroyFuncs.forEach((f) => f());
    destroyFuncs.length = 0;
  });
  const propsCopy = {};
  Object.keys(props).forEach((key) => propsCopy[key] = props[key]);
  solidJs.createEffect(() => {
    const keys = Object.keys(props);
    const changes = {};
    let changesExist = false;
    keys.forEach((key) => {
      const currentValue = props[key];
      const previousValue = propsCopy[key];
      if (previousValue !== currentValue) {
        changes[key] = currentValue;
        propsCopy[key] = currentValue;
        changesExist = true;
      }
    });
    if (changesExist) {
      agGridCommunity.ComponentUtil.processOnChange(changes, api);
    }
  });
  solidJs.onMount(() => {
    const modules = props.modules || [];
    const portalManager = {
      addPortal: (info) => {
        setPortals([...getPortals(), info]);
      },
      removePortal: (info) => {
        setPortals(getPortals().filter((item) => item != info));
      }
    };
    const gridParams = {
      providedBeanInstances: {
        frameworkComponentWrapper: new SolidCompWrapperFactory(portalManager)
      },
      modules,
      frameworkOverrides: new SolidFrameworkOverrides()
    };
    const gridOptions = agGridCommunity.ComponentUtil.combineAttributesAndGridOptions(props.gridOptions, props);
    const createUiCallback = (context2) => {
      setContext(context2);
      const ctrlsService = context2.getBean(agGridCommunity.CtrlsService.NAME);
      ctrlsService.whenReady(() => {
        const refCallback = props.ref && props.ref;
        if (refCallback) {
          const gridRef = {
            api,
            columnApi: new agGridCommunity.ColumnApi(api)
          };
          refCallback(gridRef);
        }
        destroyFuncs.push(() => api.destroy());
      });
    };
    const acceptChangesCallback = () => {
    };
    const gridCoreCreator = new agGridCommunity.GridCoreCreator();
    api = gridCoreCreator.create(
      // @ts-ignore
      eGui,
      gridOptions,
      createUiCallback,
      acceptChangesCallback,
      gridParams
    );
  });
  return web.ssr(_tmpl$17, "height:100%", context() && web.escape(web.createComponent(gridComp_default, {
    get ["class"]() {
      return props.class;
    },
    get context() {
      return context();
    }
  })), web.escape(web.createComponent(solidJs.For, {
    get each() {
      return getPortals();
    },
    children: (info, i) => web.createComponent(web.Portal, {
      get mount() {
        return info.mount;
      },
      get children() {
        return web.createComponent(info.SolidClass, web.mergeProps(() => info.props));
      }
    })
  })));
};
var agGridSolid_default = AgGridSolid;

// src/index.tsx
exports.default = agGridSolid_default;

module.exports = exports.default;
